gap/generic/kernel.gi

```diff
@@ -16,29 +16,99 @@
         Assert(2, ValidateHomomInput(ri, x));
-        s := SLPforElement(rifac,ImageElm( Homom(ri), x!.el ));
+        s := SLPforElement(ImageRecogNode(ri),
+                                 ImageElm(Homom(ri), x!.el));
         if s = fail then
-            return false;
+            return fail;
         fi;
         y := ResultOfStraightLineProgram(s, ri!.pregensfacwithmem);
-        Add(l,x^-1*y);
-        if InfoLevel(InfoRecog) >= 2 then
-            Print(".\c");
+        z := x^-1*y;
+        if isone(ri)(z) or ForAny(gens, x -> isequal(ri)(x, z)) then
+            continue;
+        fi;
+        if not doVerification or SLPforElement(KernelRecogNode(ri), z!.el) = fail then
+            Add(gens, z);
+            verificationSuccess := not doVerification;
         fi;
     od;
-    if InfoLevel(InfoRecog) >= 2 then
-        Print("\n");
+    return verificationSuccess;
+  end );
+
+InstallGlobalFunction( ImmediateVerification,
+  function(ri)
+    local verified;
+    verified := GenerateRandomKernelElementsAndOptionallyVerifyThem(
+        ri,
+        RECOG_NrElementsInImmediateVerification,
+        true
+    );
+    if verified = fail then
+        ErrorNoReturn("Very bad: image was wrongly recognised ",
+                      "and  we found out too late");
     fi;
-    return true;
+    if verified = true then return true; fi;
+    # Now, verified = false.
+    Info(InfoRecog,2,
+         "Immediate verification: found extra kernel element(s)!");
+    if FindKernelFastNormalClosure(ri,5,5) = fail then
+        ErrorNoReturn("Very bad: image was wrongly recognised ",
+                      "and  we found out too late");
+    fi;
+    Info(InfoRecog,2,"Have now ",Length(gensN(ri)),
+         " generators for kernel.");
+    return false;
+  end );
+
+InstallGlobalFunction( FindKernelRandom,
+  function(ri,n)
+    local res;
+    Info(InfoRecog,2,"Creating ",n," random generators for kernel.");
+    # We need to be careful with creating trivial kernels because we may want
+    # to do immediate verification. If we find only trivial generators for the
+    # kernel, then the kernel node of `ri` is set to `fail`. However, then
+    # doing immediate verification would become a bit complicated. On the other
+    # hand, creating a kernel node with only a trivial generator breaks all
+    # kinds of other functions. So instead, when we find only trivial
+    # generators we generate more random kernel elements. This is equivalent
+    # to doing one immediate verification.
+    res := GenerateRandomKernelElementsAndOptionallyVerifyThem(ri, n, false);
+    if res = fail then return fail; fi;
+    if IsEmpty(gensN(ri)) and immediateverification(ri) then
+        Info(InfoRecog,2,"Found only trivial generators for the kernel. ",
+             "Doing immediate verification.");
+        res := GenerateRandomKernelElementsAndOptionallyVerifyThem(
+            ri,
+            RECOG_NrElementsInImmediateVerification,
+            false
```

Comment:
Confusing: you are "doing immediate verification" but here `doVerification` is false -- is this a bug? Or a case where there are two (or more) different meanings of "verification"?

Reply:
I am confused too. I don't understand why `doVerification` is set to false.

Reply:
Thanks for the feedback! This is confusing indeed, but not a bug. I misplaced the comment explaining this, it's a few lines further up. Also, I'll try to improve the comment. There definitely is room for improvement. :)

Reply:
Hmm, now that I think about it, we probably don't need this special treatment anymore... I introduced that before I added the fix for a bug in `SLPforElementGeneric`. Now that I have that fix, the change we're talking about can probably be removed. I'll run the test suite and see.

Reply:
I disabled this case, ran the test suite, and it failed. So I'll just keep this in.

Reply:
Ok, the comment is now easier to understand.